10_rest/willshen8/pkg/puppy/rest.go

```diff
@@ -0,0 +1,78 @@
+package puppy
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/render"
+)
+
+// RestHandler provides a puppy storer to store the puppies
+type RestHandler struct {
+	// sync.Mutex
+	store Storer
+}
+
+// NewRestHandler is a factory method that makes new storer
+func NewRestHandler(storer Storer) *RestHandler {
+	return &RestHandler{
+		store: storer,
+	}
+}
+
+// HandleGet gets the puppy by id and displays the results.
+func (rh *RestHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
+	id, _ := strconv.Atoi(chi.URLParam(r, "id"))
```

Comment:
what if I make req like `api/puppy/foo -X GET`.
I may be wrong, but `id, _ = 0, someErr` and you will continue puppy read with id 0

Reply:
you get this message: 
```{"message":"Puppy ID can not be found, read operation failed.","code":1002}
```

Reply:
And this is wrong in my mind.

Please consider the given example:
```
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10 ●   curl localhost:8888/api/puppy/0 -I -X GET
HTTP/1.1 200 OK <------------------- HERE OK
Content-Type: application/json; charset=utf-8
Date: Fri, 16 Aug 2019 14:49:58 GMT
Content-Length: 76

(⎈ N/A:N/A)
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10 ●   curl localhost:8888/api/puppy/foo -I -X GET
HTTP/1.1 200 OK <-------------- HERE Not OK
Content-Type: application/json; charset=utf-8
Date: Fri, 16 Aug 2019 14:50:02 GMT
Content-Length: 76

(⎈ N/A:N/A)

```

I would expect `400` as a response for `api/puppy/foo`. This is bad request in my mind.
I'd also expect a different error message like: `param id should be type int` or sth similar.

I'm getting the response you posted because you are indexing puppies from 1 and puppy with id: 0 doesn't exist. This gives me `puppy not found` error for the bad request (`foo` is not valid `id`)

Reply:
I see what you mean, so my code now will return bad request error message instead of id not found. Thanks

Reply:
Great! you solved the biggest issue!

I don't want to be a pain on the neck, but please also be aware of HTTP codes. due to RFC(https://tools.ietf.org/html/rfc7231#section-6.5.1) server should return proper code in the header.

Response 200 means, everything is ok and given respone contains some valid data, not error description. I know some pages ignores this :(

In this case, I believe we're dealing with error, not a success. you can server error in a proper way, but it's still an error.

Please check snipset below. Response code is 200 and in description you've got 400.
```
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10 ●   curl localhost:8888/api/puppy/foo -X GET -v
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8888 (#0)
> GET /api/puppy/foo HTTP/1.1
> Host: localhost:8888
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 200 OK <----------------------------------HERE
< Content-Type: application/json; charset=utf-8
< Date: Sat, 17 Aug 2019 11:34:53 GMT
< Content-Length: 83
<
{"message":"Invalid puppy ID, ID must be numerical and greater than 0","code":400} <---------HERE
* Connection #0 to host localhost left intact
```

As a proper example for `NotFound`
```
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10 ●   curl example.com/no-such-page -I
HTTP/1.1 404 Not Found <------ HERE - Error code
Accept-Ranges: bytes
Cache-Control: max-age=604800
Content-Type: text/html; charset=UTF-8
Date: Sat, 17 Aug 2019 12:02:39 GMT
Expires: Sat, 24 Aug 2019 12:02:39 GMT
Last-Modified: Sat, 17 Aug 2019 12:02:30 GMT
Server: ECS (sjc/4E32)
X-Cache: 404-HIT
Content-Length: 1270
```
In your case it's `200` + proper response content
```
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8888 (#0)
> GET /api/puppy/33 HTTP/1.1
> Host: localhost:8888
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 200 OK <------- HERE OK
< Content-Type: application/json; charset=utf-8
< Date: Sat, 17 Aug 2019 12:07:09 GMT
< Content-Length: 76
<
{"message":"Puppy ID can not be found, read operation failed.","code":1002}
* Connection #0 to host localhost left intact
```

This issue exists in other endpoints too.

Reply:
Thanks, it has been fixed.
```2019/08/18 15:05:41 "GET http://localhost:8888/api/puppy/100 HTTP/1.1" from [::1]:58995 - 400 76B in 
148.952µs
2019/08/18 15:05:48 "GET http://localhost:8888/api/puppy/foo HTTP/1.1" from [::1]:58995 - 400 83B in 
35.927µs
```